src/Cookie.php

```diff
@@ -0,0 +1,494 @@
+    protected $maxAge;
+
+    /**
+     * @var \DateTime|null
+     */
+    protected $expires;
+
+    /**
+     * @var string|null
+     */
+    protected $domain;
+
+    /**
+     * @var string
+     */
+    protected $path;
+
+    /**
+     * @var bool
+     */
+    protected $secure;
+
+    /**
+     * @var bool
+     */
+    protected $httpOnly;
+
+    /**
+     * @param string             $name
+     * @param string|null        $value
+     * @param int|\DateTime|null $expiration
+     * @param string|null        $domain
+     * @param string|null        $path
+     * @param bool               $secure
+     * @param bool               $httpOnly
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(
+        $name,
+        $value = null,
+        $expiration = 0,
+        $domain = null,
+        $path = null,
+        $secure = false,
+        $httpOnly = false
+    ) {
+        $this->validateName($name);
+        $this->validateValue($value);
+
+        $this->name = $name;
+        $this->value = $value;
+        $this->maxAge = is_int($expiration) ? $expiration : null;
+        $this->expires = $this->normalizeExpires($expiration);
+        $this->domain = $this->normalizeDomain($domain);
+        $this->path = $this->normalizePath($path);
+        $this->secure = (bool) $secure;
+        $this->httpOnly = (bool) $httpOnly;
+    }
+
+    /**
+     * Returns the name.
+     *
+     * @return string
+     */
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    /**
+     * Returns the value.
+     *
+     * @return string|null
+     */
+    public function getValue()
+    {
+        return $this->value;
+    }
+
+    /**
+     * Checks if there is a value.
+     *
+     * @return bool
+     */
+    public function hasValue()
+    {
+        return isset($this->value);
+    }
+
+    /**
+     * Sets the value.
+     *
+     * @param string|null $value
+     *
+     * @return self
+     */
+    public function withValue($value)
+    {
+        $this->validateValue($value);
+
+        $new = clone $this;
+        $new->value = $value;
+
+        return $new;
+    }
+
+    /**
+     * Returns the max age.
+     *
+     * @return int|null
+     */
+    public function getMaxAge()
+    {
+        return $this->maxAge;
+    }
+
+    /**
+     * Checks if there is a max age.
+     *
+     * @return bool
+     */
+    public function hasMaxAge()
+    {
+        return isset($this->maxAge);
+    }
+
+    /**
+     * Sets the max age.
+     *
+     * @param int|null $maxAge
+     *
+     * @return self
+     */
+    public function withMaxAge($maxAge)
+    {
+        $new = clone $this;
+        $new->maxAge = is_int($maxAge) ? $maxAge : null;
+
+        return $new;
+    }
+
+    /**
+     * Sets both the max age and the expires attributes.
+     *
+     * @param int|\DateTime|null $expiration
+     *
+     * @return self
+     */
+    public function withExpiration($expiration)
```

Comment:
expires is the old way of doing things, HTTP 1.1 has max-age: http://mrcoles.com/blog/cookies-max-age-vs-expires/ (IE9+ supports max-age). i guess we can keep expires support, but i would not support the mixed approach. if anything, it would be the job of some hypothetical server side code that sets expires from the max-age.

though when we speak about client side code, we could receive a cookie with both values set and might care about that (and be it just to report that the server is sending ambigous data) - so probably setting one value should not null out the other. 

but i would still remove this method, it seems weird. and we could phpdoc that expires is the old way of doing things and no longer recomended